src/server/qgsfcgiserverrequest.cpp

```diff
@@ -173,35 +173,49 @@ void QgsFcgiServerRequest::readData()
   if ( lengthstr )
   {
     bool success = false;
-    int length = QString( lengthstr ).toInt( &success );
-    // Note: REQUEST_BODY is not part of CGI standard, and it is not
-    // normally passed by any CGI web server and it is implemented only
-    // to allow unit tests to inject a request body and simulate a POST
-    // request
-    const char *request_body  = getenv( "REQUEST_BODY" );
-    if ( success && request_body )
+    const int length = QString( lengthstr ).toInt( &success );
+    if ( !success || length < 0 )
     {
-      QString body( request_body );
-      body.truncate( length );
-      mData.append( body.toUtf8() );
-      length = 0;
+      QgsMessageLog::logMessage( "fcgi: Invalid CONTENT_LENGTH",
+                                 QStringLiteral( "Server" ), Qgis::MessageLevel::Critical );
+      mHasError = true;
     }
+    else
+    {
+      // Note: REQUEST_BODY is not part of CGI standard, and it is not
+      // normally passed by any CGI web server and it is implemented only
+      // to allow unit tests to inject a request body and simulate a POST
+      // request
+      const char *request_body = getenv( "REQUEST_BODY" );
+
 #ifdef QGISDEBUG
-    qDebug() << "fcgi: reading " << lengthstr << " bytes from " << ( request_body ? "REQUEST_BODY" : "stdin" );
+      qDebug() << "fcgi: reading " << lengthstr << " bytes from " << ( request_body ? "REQUEST_BODY" : "stdin" );
 #endif
-    if ( success )
-    {
-      // XXX This not efficient at all  !!
-      for ( int i = 0; i < length; ++i )
+
+      if ( request_body )
       {
-        mData.append( getchar() );
+        const size_t request_body_length = strlen( request_body );
+        const int actual_length = static_cast<int>( std::min<size_t>( length, request_body_length ) );
```

Comment:
I realise it's matching some other formatting used here, but could we avoid the non-camel case creep for the new variables?

Reply:
fixed